diff -uNr src/easy-rsa/EasyRSA-2.2.2/openssl.cnf src-new/easy-rsa/EasyRSA-2.2.2/openssl.cnf
--- src/easy-rsa/EasyRSA-2.2.2/openssl.cnf	2018-04-09 17:13:27.000000000 +0800
+++ src-new/easy-rsa/EasyRSA-2.2.2/openssl.cnf	2018-04-09 17:14:38.000000000 +0800
@@ -54,9 +54,9 @@
 # so this is commented out by default to leave a V1 CRL.
 # crl_extensions	= crl_ext
 
-default_days	= 3650			# how long to certify for
+default_days	= 7300			# how long to certify for
 default_crl_days= 30			# how long before next CRL
-default_md	= md5			# which md to use.
+default_md	= sha256			# which md to use.
 preserve	= no			# keep passed DN ordering
 
 # A few difference way of specifying how similar the request should look
@@ -89,7 +89,7 @@
 [ req ]
 default_bits		= $ENV::KEY_SIZE
 default_keyfile 	= privkey.pem
-default_md		= md5
+default_md		= sha256
 distinguished_name	= req_distinguished_name
 attributes		= req_attributes
 x509_extensions	= v3_ca	# The extentions to add to the self signed cert
